File: java/src/DisjointSet.java
import java.util.HashMap;
import java.util.Map;

public class DisjointSet<T> {

    private final Map<T, T> parent = new HashMap<>();
    private final Map<T, Integer> size = new HashMap<>();
    private int count = 0;

    public void insert(T x) {
        if (parent.get(x) != null) return;

        parent.put(x, x);
        size.put(x, 1);
        count++;
    }

    public T root(T x) {
        var r = x;
        var s = parent.get(r);

        if (s == null)
            throw new IllegalArgumentException();

        while (r != s) {
            r = s;
            s = parent.get(s);
        }

        return r;
    }

    public boolean isConnected(T a, T b) {
        return a == b || root(a) == root(b);
    }

    public void union(T a, T b) {
        if (a == b) return;

        var rootA = root(a);
        var rootB = root(b);

        if (rootA == rootB) return;

        var sizeA = size.get(rootA);
        var sizeB = size.get(rootB);

        if (sizeA < sizeB) {
            parent.put(rootA, rootB);
            size.put(rootB, sizeA + sizeB);
        }
        else {
            parent.put(rootB, rootA);
            size.put(rootA, sizeA + sizeB);
        }

        count--;
    }

    public int count() {
        return count;
    }
}

File: c/disjoint_set.c
#include <stdlib.h>

typedef struct {
  int *parent;
  int *weight;
} disjoint_set;

disjoint_set create(int size) {
  disjoint_set set;

  set.parent = malloc(sizeof(int) * size);
  set.weight = malloc(sizeof(int) * size);

  for (int i = 0; i < size; i++) {
    set.parent[i] = i;
    set.weight[i] = 1;
  }

  return set;
}

int root(disjoint_set set, int index) {
  int old_root = index;
  int root = set.parent[index];

  while (old_root != root) {
    old_root = root;
    root = set.parent[root];
  }

  return root;
}

void union_elements(disjoint_set set, int a, int b) {
  if (a == b) return;

  int root_a = root(set, a);
  int root_b = root(set, b);

  if (root_a == root_b) return;

  int weight_a = set.weight[root_a];
  int weight_b = set.weight[root_b];

  if (weight_a < weight_b) {
    set.parent[root_a] = root_b;
    set.weight[root_b] += weight_a;
  } else {
    set.parent[root_b] = root_a;
    set.weight[root_a] += weight_b;
  }
};

void destroy(disjoint_set set) {
  free(set.parent);
  free(set.weight);
}
